src/core/shader.js

```diff
@@ -96,8 +98,32 @@ Shader.prototype = {
     var schema = this.schema;
     dataKeys.forEach(processData);
     function processData (key) {
+      var varKey;
+      var materialKey;
       if (!schema[key] || schema[key].is !== type) { return; }
-      if (variables[key].value === data[key]) { return; }
+      if (schema[key].type === 'map') {
+        // Special handling is needed for textures.
+        // For textures, by convention data.src becomes uniform sampler2D map.
+        if (key === 'src' && type === 'uniform') {
+          varKey = 'map';
+          materialKey = 'map';
+        } else {
+          varKey = key;
+          materialKey = '_texture_' + key;
+        }
+        // If data unchanged, get out early.
+        if (!variables[varKey] || variables[varKey].value === data[key]) { return; }
+        // We can't actually set the variable correctly until we've loaded the texture.
+        self.el.addEventListener('materialtextureloaded', function (e) {
```

Comment:
We don't have to refactor here, but it'll make it simpler and easy to reason later if we try a straight up TextureLoader here rather than jumping into another pile of code.